torch/optim/adadelta.py

```diff
@@ -43,6 +43,8 @@ def __init__(
         )
         super().__init__(params, defaults)
 
+    # See #123400
+    @torch._disable_dynamo
```

Comment:
Looking this up in the codebase, most of the usages of torch._disable_dynamo are in optimizer code today. There are also a few in test cases, checkpointing, internal dynamo code, but is this the right way to disable dynamo? Is this the same thing as a graph break? What's the difference between this decorator and the torch patcher? Is there other precedent for disabling dynamo on library code?

Reply:
This is the correct way to disable dynamo. The patcher is more indirect - having this decorator here is clearer, otherwise there is no difference. Graph breaks only avoid tracing a piece of code and resume afterward, disabling will disable dynamo on the entirety of a function and run it in python.

cc @anijain2305

Reply:
Yes, this is the right way to disable dynamo. The torch patcher is a remnant of codebase when TorchDynamo used to live outside of pytorch repo. Directly applying decorator ensures that its modified inplace, and anybody else reading from the source file will get the disabled function.

disable is different and stronger than a graph break. disable means Dynamo don't intercept on any of the frames from this function and its child functions. graph break is just a break in the function - the code before and after the break still goes through dynamo.

Reply:
Okay, even if this is the right technical approach, I am still unsure this is the direction we wanna go--why is disabling the right thing to do here? What other alternatives do we have? Is there other precedent for disabling dynamo on library code outside of optimizers?

Reply:
I think .. it depends on the priority.

If something is hard to support with torch.compile + not popular, then it brings the question - is this high priority? If not, one of the compromises is to add `disable` so that even if some rare user does use it, it does not lead to wrong results. `disable` will make it fallback to `eager`, so torch.compile will be as good as eager for this function.

We have applied `_disable_dynamo` (lazy version of `disable`) at many places in torch codebase (like don't trace inside a torch-scripted model, or don't trace inside the checkpoint API because we don't support saved_tensor_hooks).

So the question is - is this usecase very popular? I have near-zero experience with optimizers. So, I will let @janeyx99 and @mlazos decide on that.

Reply:
Basically my take here is that this is loading a saved optimizer checkpoint, this will rarely occur in a model's compiled region if ever and there's zero benefit to compiling this. If we want to investigate this separately that's fine, but I think this is a valid solution to fix failures in the dynamo-traced optimizer tests and shouldn't be blocking.

Reply:
How does the nn module's equivalent `__setstate__` in module.py get handled in dynamo? How come we don't disable that in dynamo? Or is it that no test has tested this so dynamo hasn't needed to be considered for setting state for modules?

I am curious if this is a more general state API topic vs something particular to optimizers.

Reply:
Figured out a way to remove the disable, should be good now